Wayland displaying using shared memory between virtual machines

jkuro-tii opened the pull request:
<!--
    Copyright 2023 TII (SSRC) and the Ghaf contributors
    SPDX-License-Identifier: CC-BY-SA-4.0
-->

## Description of changes
Add memory sharing mechanism between virtual machines and use it for wayland/waypipe displaying.

#### Main changes:

### Host

- Added the shared memory manager within QEMU.
- Direct Memory Mapping via SysBus:
  - Modified QEMU's code to map shared memory directly into the VM's memory address space using the SysBus interface.
- Option for Defining Shared Memory Manager:
  - Added an option to QEMU for defining a shared memory manager when running VMs.
- BigPages Area Allocation:
  - Allocated shared memory within the BigPages area for optimized memory usage.
 
### GUI and app VMs

1. **Linux Driver for Shared Memory and Interrupts:**
   - Added a Linux driver to manage shared memory operations and interrupts and creating the `/dev/ivshmem` device.

2. **Memsocket Application for Inter-VM Communication:**
   - Introduced the `memsocket` application to facilitate socket data forwarding between VMs.

3. **Integration with Waypipe:**
   - Modified Waypipe service options to interface with the `memsocket` application, enabling seamless communication between VMs.

<!--
Summary of the proposed changes in the PR description in your own words. For dependency updates, please link to the changelog.
-->

## Checklist for things done

<!-- Please check, [X], to all that applies. Leave [ ] if an item does not apply but you have considered the check list item. Note that all of these are not hard requirements. They serve information to reviewers. When you fill the checklist, you indicate to reviewers you appreciate their work. -->

- [x] Summary of the proposed changes in the PR description
- [x] More detailed description in the commit message(s)
- [x] Commits are squashed into relevant entities - avoid a lot of minimal dev time commits in the PR
- [x] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [ ] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [x] PR linked to architecture documentation and requirement(s) [ticket id](https://ssrc.atlassian.net/browse/SP-3805)
- [x] Test procedure described (or includes tests). Select one or more:
  - [x] Tested on Lenovo X1 `x86_64`
  - [ ] Tested on Jetson Orin NX or AGX `aarch64`
  - [ ] Tested on Polarfire `riscv64`
- [ ] Author has run `nix flake check --accept-flake-config` and it passes
- [ ] All automatic Github Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [x] Author has added reviewers and removed PR draft status

<!-- Additional description of omitted [ ] items if not obvious. -->

## Testing
Run Chromium and other wayland enabled application. Verify that they execute and display properly.
Run performance tests described in [documentation](https://ssrc.atlassian.net/wiki/spaces/~62552e6ffdb60b006927ad98/pages/825720835/Wayland+displaying+with+shared+memory)
<!--
How this was tested by the author? How is this supposed to be tested
by people doing system testing?
-->

## Jira ticket:
https://ssrc.atlassian.net/browse/SP-3805

## Documentation:
https://ssrc.atlassian.net/wiki/spaces/~62552e6ffdb60b006927ad98/blog/2022/09/29/612958326/Memory+sharing+between+virtual+machines
https://ssrc.atlassian.net/wiki/spaces/~62552e6ffdb60b006927ad98/pages/825720835/Wayland+displaying+with+shared+memory

## Summary of performance
Playing YouTube is smooth, up to a resolution of 2160p. 
CPU consumption by the memsocket application is below 5%, usually 3-6%.
iperf results are presented in [documentation](https://ssrc.atlassian.net/wiki/spaces/~62552e6ffdb60b006927ad98/pages/825720835/Wayland+displaying+with+shared+memory)

jenninikko commented:
Are you planning to add documentation to the repository?

vilvo reviewed (changes requested):
1. Please make PR description according to the template - https://github.com/tiiuae/ghaf/blob/main/.github/pull_request_template.md 
2. Please fill the PR template checklist, describe testing and remove "Draft" status.
3. Re-request review

jkuro-tii commented:
The purpose of this PR is to share the code and try together to solve the mentioned problem of high kernel overhead.

jkuro-tii commented:
Getting pert data. perf is a standard Linux tool (https://perf.wiki.kernel.org/index.php/Main_Page). In this PR the ``perf`` component is included in the host, GuiVM and AppVM VMs. The first stage of using ``perf`` is collecting execution data (e.g. ``sudo perf record -a -g -p PID``) for certain period of time. The data is written into the ``perf.data`` file. Then the data can be visualized using the ``sudo perf report`` subcommand. ``perf report`` uses ``/proc/kallsyms`` for obtaining information about kernel symbols. If you want to process the data offline, copy ``/proc/kallsyms`` and refer to it with the ``perf record`` ``--kallsyms=<file>`` option. The files that I attached contain both ``perf.data`` and kernel symbols files.
 
To check the memsocket application performace on GuiVM:

1. Find the PID of the memsocket service: ``systemctl status --user memsocket.service | grep "Main PID"``
2. Collect ``perf`` data: ``sudo perf record -a -g -p PID``
3. Process it with ``perf report``.

To check GuiVM's qemu performance on the host:
1. Start terminal
2. Login to host: ``ssh 192.168.101.2``
3. Find PID of qemu instance running GuiVM: ``systemctl status microvm@gui-vm.service | grep "Main PID"``
4. collect perf data: ``sudo perf record -a -g -p PID``
5. Process it with ``perf report``.

All files kept in the ``/home/ghaf`` directory on the host are stored on the Ghaf boot disk. It may be used for accessing them on another computer. E.g. files from the GuiVM machine can be copied to host using ``scp _file_ ghaf@192.168.101.2:`` and accessed offline.
Sample reading about ``perf``:
[1](https://linux-kernel-labs.github.io/refs/heads/master/labs/kernel_profiling.html), [2](https://perf.wiki.kernel.org/index.php/Tutorial)

vilvo commented:
> To check the memsocket application performace on GuiVM

Thanks. I can reproduce your `perf` measurement and `perf report` analysis in the Youtube playing scenario. In addition, the user experience is terrible.

[Behind the link of the description](https://ssrc.atlassian.net/wiki/spaces/~62552e6ffdb60b006927ad98/pages/825720835/Wayland+displaying+with+shared+memory), there an interesting mention on 
```
Result
Throughput socket to socket: 
2.8 GB/s = 22.4 Gb/s
using ssh: 140 MB/s = 1.1 Gb/s

Lenovo Thinkpad P1 Intel i7 @2.5 GHz

Sending 1 MiB buffer 8 K times (total 80 GiB sent)
``` 

Can you please share reproducible `iperf` test description? I'd like to reproduce and establish Lenovo X1 Carbon Gen11 baseline throughput. Another thing we could then check is to parametrize the buffer size and find where we hit the ceiling.

jkuro-tii commented:
I did iperf measuring, with the unsock tool - in order to tunnel TCP traffic over socket and shared memory. 

Test configuration:
_chromium-vm_: `iperf -c` with `unsock` -> [socket] -> `memsocket`-> 
[shared memory]
 _guivm_: `memsocket` -> [socket] -> `iperf -s` with `unsock`

Performance achieved:
Lenovo X1 with Ghaf: ~90 Mbits/sec
Lenovo P1 with Fedora Linux and two Ghaf vm-debug machines (my development environment): ~32 **Gbit**/s


[result.client_x1.log](https://github.com/tiiuae/ghaf/files/14308081/result.client_x1.log)
[result.server_x1.log](https://github.com/tiiuae/ghaf/files/14308082/result.server_x1.log)
[result.server.log](https://github.com/tiiuae/ghaf/files/14308083/result.server.log)
[result.client.log](https://github.com/tiiuae/ghaf/files/14308084/result.client.log)

vilvo commented:
> I did iperf measuring, with the unsock tool - in order to tunnel TCP traffic over socket and shared memory.
> 
> Test configuration: _chromium-vm_: `iperf -c` with `unsock` -> [socket] -> `memsocket`-> [shared memory] _guivm_: `memsocket` -> [socket] -> `iperf -s` with `unsock`
> 
> Performance achieved: Lenovo X1 with Ghaf: ~90 Mbits/sec Lenovo P1 with Fedora Linux and two Ghaf vm-debug machines (my development environment): ~32 **Gbit**/s
> 
> [result.client_x1.log](https://github.com/tiiuae/ghaf/files/14308081/result.client_x1.log) [result.server_x1.log](https://github.com/tiiuae/ghaf/files/14308082/result.server_x1.log) [result.server.log]

>(https://github.com/tiiuae/ghaf/files/14308083/result.server.log) [result.client.log](https://github.com/tiiuae/ghaf/files/14308084/result.client.log)

```
[ghaf@ghaf-host:~/shmsockproxy/app]$ UNSOCK_ADDR=127.0.0.1 UNSOCK_DIR=`pwd` LD_P
RELOAD=~/unsock/libunsock.so.1.1.0  iperf -c 127.0.0.1 -p 1234
```
This indicates you have compiled some dependency in the current working directory (`pwd`?) of shmsockproxy app to use some version(which git commit hash?) of `libunsock` with iperf. From above instructions it's unclear on how to replicate the test setup as I rather not do guessing.

Optionally the test tools you have used could be nix packaged and included in the PR `-debug` build.

Would it make sense to add some simple unit throughput test through memsocket instead? What do you think?

vilvo commented:
If I try a bit further, to compile the `memsocket` on `ghaf-host` I get a lot of gcc warnings on `-Wformat-extra-args` and `-Wformat-overflow=` (just a small snippet below).

```
[ghaf@ghaf-host:~]$ nix-shell -p git stdenv

[nix-shell:~]$ cd shmsockproxy/app/

[nix-shell:~/shmsockproxy/app]$ gcc memsocket.c
memsocket.c: In function ‘server_init’:
memsocket.c:175:8: warning: too many arguments for format [-Wformat-extra-args]
  175 |   INFO("server initialized", "");
...
memsocket.c: In function ‘wayland_connect’:
memsocket.c:50:19: warning: ‘%s’ directive writing up to 255 bytes into a region of size 235 [-Wformat-overflow=]
...
```

vilvo commented:
Also, I think this is relevant to performance comparison between X1 and desktop - https://github.com/tiiuae/ghaf/pull/340#pullrequestreview-1884426193
So we can now use Ghaf internal NVMe instead of external SSD over USB3.

jkuro-tii commented:
chromium -> [socket] -> waypipe  -> [socket] -> memsocket -> 

[shared memory]

-> memsocket send() -> [socket] -> waypipe -> wayland

> If I try a bit further, to compile the `memsocket` on `ghaf-host` I get a lot of gcc warnings on `-Wformat-extra-args` and `-Wformat-overflow=` (just a small snippet below).
> 
> ```
> [ghaf@ghaf-host:~]$ nix-shell -p git stdenv
> 
> [nix-shell:~]$ cd shmsockproxy/app/
> 
> [nix-shell:~/shmsockproxy/app]$ gcc memsocket.c
> memsocket.c: In function ‘server_init’:
> memsocket.c:175:8: warning: too many arguments for format [-Wformat-extra-args]
>   175 |   INFO("server initialized", "");
> ...
> memsocket.c: In function ‘wayland_connect’:
> memsocket.c:50:19: warning: ‘%s’ directive writing up to 255 bytes into a region of size 235 [-Wformat-overflow=]
> ...
> ```
> 
> Edit. `git` was used to clone:
> 
> ```
> [nix-shell:~]$ git clone https://github.com/tiiuae/shmsockproxy.git
> [nix-shell:~]$ cd shmsockproxy/
> [nix-shell:~/shmsockproxy]$ git log --oneline -1
> 8b1bf60 (HEAD -> main, origin/main, origin/HEAD) Removed forking, as systemd sends SIGTERM
> ```

I added binary library and a few commands by hand into chromium and gui VMs. Is it OK it I attach them here, or you want to include them in the Ghaf build?

vilvo commented:
> I added binary library and a few commands by hand into chromium and gui VMs. Is it OK it I attach them here, or you want to include them in the Ghaf build?

This command:
```
[ghaf@ghaf-host:~/shmsockproxy/app]$ UNSOCK_ADDR=127.0.0.1 UNSOCK_DIR=`pwd` LD_P
RELOAD=~/unsock/libunsock.so.1.1.0  iperf -c 127.0.0.1 -p 1234
```
from [here](https://github.com/tiiuae/ghaf/files/14308084/result.client.log) indicates that some commands are also required on `ghaf-host` instead of only `chromium-vm` and `gui-vm` _if one wants to reproduce the iperf throughput measurements._ Now I'm quite not sure if that's the case or not. So I still have some trouble following the testing procedure and if we are talking about the integrated YouTube-streaming case or the iperf measurement case. So it's probably more clear if we add reproducible test setup in the Ghaf build for `-debug`-target and clarify instructions.

jkuro-tii commented:
It doesn't seem we are facing data loss. Connection between remote and waypipe instances is multichannel, data is sent both directions. After portion of data sent from AppVM, GuiVM sends a short package of data, probably it's confirmation. If its not received by AppVM, there is no further transmission. Also if we faced data loss, GUI stuff - windows, menu, scroll bars would be damaged. We don't face it.

I've pushed my branch rebased with main.

vilvo commented:
> I've pushed my branch rebased with main.

Thanks, I installed your rebased draft branch build to X1 NVMe with:
```
nix build .#packages.x86_64-linux.lenovo-x1-carbon-gen11-debug-installer
dd if=result/iso/nixos-23.11.20231218.3a9928d-x86_64-linux.iso ...
<boot and install to nvme0n1>
```

The problem with `dd` method is the numbers tell very little. Writing stops with
```
[ghaf@gui-vm:~]$ time dd if=/dev/zero of=/dev/ivshmem
dd: writing to '/dev/ivshmem': No space left on device
32769+0 records in
32768+0 records out
16777216 bytes (17 MB, 16 MiB) copied, 0.688331 s, 24.4 MB/s

real	0m0.692s
user	0m0.000s
sys	0m0.690s
```

vilvo commented:
During the Youtube streaming scenario there's very heavy load in QEMU processes as seen from `htop` on `ghaf-host`.
The message signaled interrupts on `chromium-vm` and `gui-vm` are roughly ~50-60/s as seen with `watch -n1 "cat /proc/interrupts"`
![qemu-load-with-youtube-streaming](https://github.com/tiiuae/ghaf/assets/1027150/87101ecc-7e63-48e0-9742-b5f0f2476855)
When you stop the streaming, there's no  QEMU load.

vilvo commented:
We're not using `aio=native` with QEMU. Please check
https://hex.ro/wp/blog/kvm-qemu-high-cpu-load-but-low-cpu-usage/
I checked the `strace` during streaming and we've got the same - littered with `events=POLLIN` and a ton of `anon_inode:[eventfd]`s under the QEMU `/proc/<pid>/fd`

vilvo commented:
And this presentation in the blog post link of the previous comment is extremely relevant for optimizing our case - https://vmsplice.net/~stefan/stefanha-kvm-forum-2017.pdf

jkuro-tii commented:
The ``aio`` parameter is not accepted for device that is used for shared memory and interrupts:
``                          "-device"
                          "ivshmem-doorbell,vectors=${vectors},chardev=ivs_socket,aio=native,cache=none"
                          "-chardev"
                          "socket,path=${config.ghaf.profiles.applications.ivShMemServer.hostSocketPath},id=ivs_socket"
``
:
``qemu-kvm: -device ivshmem-doorbell,vectors=10,chardev=id,aio=native: Property 'ivshmem-doorbell.aio' not found``

It seems that it's designated for disks/block devices. I'll dig deeper.

vilvo commented:
> The `aio` parameter is not accepted for device that is used for shared memory and interrupts: ` "-device" "ivshmem-doorbell,vectors=${vectors},chardev=ivs_socket,aio=native,cache=none" "-chardev" "socket,path=${config.ghaf.profiles.applications.ivShMemServer.hostSocketPath},id=ivs_socket"` : `qemu-kvm: -device ivshmem-doorbell,vectors=10,chardev=id,aio=native: Property 'ivshmem-doorbell.aio' not found`
> 
> It seems that it's designated for drives/block devices. I'll dig deeper.

aio=native was fairly new to QEMU and I think must be implemented in the device. Our ivmshmem device is based on fairly old implementation https://github.com/tiiuae/shmsockproxy/blob/main/drivers/char/ivshmem/kvm_ivshmem.c#L8
Is this something to consider?

jkuro-tii commented:
The kernel driver uses regular PCI interface. I think the matter is the counterpart implementation in [qemu](https://github.com/qemu/qemu/blob/master/hw/misc/ivshmem.c). And we still don't know why in development environment it runs without overhead.

vilvo commented:
> The kernel driver uses regular PCI interface. I think the matter is the counterpart implementation in [qemu]

Could be only qemu. Or could be both. Regarding the actual driver the differences between X1 and your desktop can be the MMU and how MSIs are handled. Please correct me if I'm wrong but my understanding is that the driver is only using logical virtio-PCI-device model, not actual PCI bus but shared memory instead. On X1 from `chromium-vm` shared memory PCI driver perspective it looks like:
```
00:07.0 RAM memory: Red Hat, Inc. Inter-VM shared memory (rev 01)
	Subsystem: Red Hat, Inc. QEMU Virtual Machine
	Control: I/O+ Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR+ FastB2B- DisINTx+
	Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
	Latency: 0
	Region 0: Memory at febfc000 (32-bit, non-prefetchable) [size=256]
	Region 1: Memory at febfd000 (32-bit, non-prefetchable) [size=4K]
	Region 2: Memory at fd000000 (64-bit, prefetchable) [size=16M]
	Capabilities: [40] MSI-X: Enable+ Count=12 Masked-
		Vector table: BAR=1 offset=00000000
		PBA: BAR=1 offset=00000800
	Kernel driver in use: kvm-shmem
	Kernel modules: virtio_pci
```

> And we still don't know why in Chromium-VM/Zathura-VM/dev environment there is no overhead.

We don't know the measured overhead in your dev environment. If your desktop CPU running the guest is not throttling, it may have observable impact but still have overhead. We also don't know the QEMU version. Please share.

jkuro-tii commented:
I designed an utility for testing connection.
On one side (e.g. chromium-vm) it sends a file to a socket, on the other side (e.g. gui-vm) it receives it, without any writing. The transmission is protected by CRC-8.
Both sockets are connected with the memsocket app.
It's possible to specify how many data to send, and to force send broken CRC (by adding anything as 4th parameter), e.g.:

``memtest socket_path /dev/random 10M``
``memtest socket_path /dev/random 10M anything # for sending broken CRC``

First start it on the receiving site (any of the VMs):
``memsocket -c ./test.socket & # run it once``
``memtest ./test.socket``

Now file(s) can be send from another VM:
``memsocket -s ./test.socket 2 & # run it once``

``memtest ./test.socket /dev/random 10M``
``memtest ./test.socket /dev/random 10M wrong_crc``

It's prints some statistics.
**Before using it it's necessary to stop the memsocket service: ``systemctl stop memsocket.service`` on the gui-vm and the other peer VM (e.g. chromium)!**

Sample sessions:

[ghaf@chromium-vm:~]$ memtest
Usage: memtest socket_path for receiving.
memtest socket_path input_file [size CRC_error] for sending.
Before using stop the memsocket service: 'systemctl stop --user memsocket.service'.
For receiving run e.g.: 'memsocket -c ./test.sock &; memtest ./test.sock'.
To send a file run on other VM: 'memsocket -s ./test.sock 3 &; memtest ./test.sock /dev/random 10M.'
To force wrong CRC on sending: 'memtest ./test.sock /dev/random 10M xxx
'
[ghaf@chromium-vm:~]$ systemctl stop --user memsocket.service

[ghaf@chromium-vm:~]$ memsocket -s ./test.sock 3 &
[1] 776

[ghaf@chromium-vm:~]$ memtest ./test.sock /dev/random 10M
Sent/received 10485761 bytes 2 Mbytes/sec
real 5.030s
user 0.010s
sys  0.030s

[ghaf@chromium-vm:~]$ memtest ./test.sock /dev/random 100M
Sent/received 104857601 bytes 4 Mbytes/sec
real 27.640s
user 0.250s
sys  0.320s
[ghaf@chromium-vm:~]$ memtest ./test.sock /dev/random 100M
Sent/received 104857601 bytes 244 Mbytes/sec
real 0.410s
user 0.200s
sys  0.200s

[ghaf@chromium-vm:~]$ memtest ./test.sock /dev/random 100M ui **# to brake the CRC**
Sent/received 104857601 bytes 263 Mbytes/sec
real 0.380s
user 0.170s
sys  0.200s

[ghaf@chromium-vm:~]$ memtest ./test.sock /dev/random 100M 
Sent/received 104857601 bytes 263 Mbytes/sec
real 0.380s
user 0.190s
sys  0.180s




--------- 
Usage: memtest socket_path for receiving.
memtest socket_path input_file [size CRC_error] for sending.
Before using stop the memsocket service: 'systemctl stop --user memsocket.service'.
For receiving run e.g.: 'memsocket -c ./test.sock &; memtest ./test.sock'.
To send a file run on other VM: 'memsocket -s ./test.sock 3 &; memtest ./test.sock /dev/random 10M.'
To force wrong CRC on sending: 'memtest ./test.sock /dev/random 10M xxx

[ghaf@zathura-vm:~]$ systemctl stop --user memsocket.service

[ghaf@zathura-vm:~]$ memsocket -c ./test.sock &
[1] 670

[ghaf@zathura-vm:~]$ memtest ./test.sock
Waiting for a connection.
Connected...
Sent/received 104857602 bytes 159 Mbytes/sec
real 0.630s
user 0.270s
sys  0.030s
Read 104857601 bytes. CRC OK
Connected...
Sent/received 104857602 bytes 244 Mbytes/sec
real 0.410s
user 0.200s
sys  0.010s
Read 104857601 bytes. CRC OK
Connected...
Sent/received 104857602 bytes 263 Mbytes/sec
real 0.380s
user 0.180s
sys  0.020s
Read 104857601 bytes. TRAMISSION ERROR!
Connected...
Sent/received 104857602 bytes 263 Mbytes/sec
real 0.380s
user 0.190s
sys  0.010s
Read 104857601 bytes. CRC OK
``

jkuro-tii commented:
I noticed that sending data between gui-vm and any other machine (zathura-vm, chromium-vm) is significantly slower then sending between app VMs (e.g. zathura-vm and chromium-vm).

On my desktop I use qemu version 8.0.5. The full run command line (it's a Ghaf vm-debug build):
``exec /nix/store/ppfkk3h94b3s3vhifp55q80kr5y8bzvc-qemu-host-cpu-only-8.0.5/bin/qemu-kvm -cpu max \
    -name ghaf-host \
    -m 1024 \
    -smp 1 \
    -device virtio-rng-pci \
    -net nic,netdev=user.0,model=virtio -netdev user,id=user.0,"$QEMU_NET_OPTS" \
    -virtfs local,path=/nix/store,security_model=none,mount_tag=nix-store \
    -virtfs local,path="${SHARED_DIR:-$TMPDIR/xchg}",security_model=none,mount_tag=shared \
    -virtfs local,path="$TMPDIR"/xchg,security_model=none,mount_tag=xchg \
    -drive cache=writeback,file="$NIX_DISK_IMAGE",id=drive1,if=none,index=1,werror=report -device virtio-blk-pci,bootindex=1,drive=drive1 \
    -device virtio-keyboard \
    -usb \
    -device usb-tablet,bus=usb-bus.0 \
    -kernel ${NIXPKGS_QEMU_KERNEL_ghaf_host:-/nix/store/sgv6vmzrbkghz8a0hr05garms07cqz2m-nixos-system-ghaf-host-23.05.20230927.5cfafa1/kernel} \
    -initrd /nix/store/sgv6vmzrbkghz8a0hr05garms07cqz2m-nixos-system-ghaf-host-23.05.20230927.5cfafa1/initrd \
    -append "$(cat /nix/store/sgv6vmzrbkghz8a0hr05garms07cqz2m-nixos-system-ghaf-host-23.05.20230927.5cfafa1/kernel-params) init=/nix/store/sgv6vmzrbkghz8a0hr05garms07cqz2m-nixos-system-ghaf-host-23.05.20230927.5cfafa1/init regInfo=/nix/store/w7b65bipqayzsi6w2ksjvd1yi7m8b0x3-closure-info/registration console=ttyS0,115200n8 console=tty0 $QEMU_KERNEL_PARAMS" \
    $QEMU_OPTS \
    "$@"
``

Mic92 commented on modules/development/debug-tools.nix:
```diff
@@ -37,6 +37,9 @@ in
           speedtest-cli
           iperf
+          perf-tools
+          linuxPackages.perf
```
Off-topic: I find bcc's profile script a bit easier to use ` sudo profile -f -F 99 10 > /tmp/folded-graph` and the result can be looked at in https://www.speedscope.app/

jkuro-tii commented:
I ran my tool in several configuration. Results:
- there is no data lost or data corruption in any scenario
- transfers between chromium-vm and zathura-vm are fast - ~250 MBytes/s
- transfers between any of the above VMs and gui-vm are very slow. They hardly reach 12 MB/s

Shutting down weston gave no result. But doing that and eliminating passing-through the GPU solved the problem. After that transfer to/from gui-vm have the same high speed as inter app-vm's.

I'll continue with different virtual PCI devices setup, and try to use qemu own monitor in order to find the bottleneck.

jkuro-tii commented:
The driver already contains interrupts counters. I verified number of interrupts sent by one VM against numbers listed in /proc/interrupts/. There are in line, i.e. number of interrupts raised by one VM is equal to number /proc/interrupts in the other VM as well as counted by the device driver. It means that there is no interrupt loop, missed interrupts.

Delays and huge CPU load happen when GPU is pass-through. Without it CPU load is less then 5%.

jkuro-tii commented:
Added some improvements:
- allocating shared memory area using huge pages (2MB size)
- adding to the qemu ivshmem driver ability to map the above memory directly in VM address space (currently it's mapped into PCI area). The physical memory address is set in kernel command line.

vilvo reviewed (changes requested):
Main observations:
- Previous PCI device model-based shared memory has been removed. I hid my outdated review comments from February.

The main issues:
- PR description field is still not following the PR template. In particular, the Testing section is both outdated (not updated) and it is still missing the exact steps to reproduce the performance testing results. The exact steps must be specific enough that people who are not familiar with the work can reproduce the measurement results.
This comment from February is still valid https://github.com/tiiuae/ghaf/pull/473#pullrequestreview-1877609249 
Please keep the PR description up-to-date.

- The PR branch does not build:
```
[vilvo@carrie:~/ghaf]$ git checkout jarek/jkuro-add-wayland-shm-lenovo_rebase
...
HEAD is now at a68cf7b Added memory sharing between VMs

[vilvo@carrie:~/ghaf]$ nix build .#packages.x86_64-linux.lenovo-x1-carbon-gen11-debug
error: hash mismatch in fixed-output derivation '/nix/store/n49a0yrl19lagg6yi2vjkcxs7mmyy55i-source.drv':
         specified: sha256-y1o9Md9apgn+mjBZ9wScWUNQ3ZgxbT0TKxl8bIMZxGs=
            got:    sha256-qNeIjJj3VJ1ngvWTa914zacQMwzpM+9KLysJyJTcnGk=
```

vilvo commented:
Structurally, these parts are copies
https://github.com/tiiuae/ghaf/pull/473/files#diff-092df22f199bcdee3f75c01eb9e440429aabe51e8fd706c44b479975f70fee72R132-R139
https://github.com/tiiuae/ghaf/pull/473/files#diff-4cd0792a1806df6e4de8a4de75cf451ec4536470b07fa5b4a3f10ad041ed989eR129-R136
so I would recommend removing the duplication.

vilvo commented:
Thanks, I tested the PR and compared 1080p streaming with main as follows:

## https://github.com/tiiuae/ghaf/pull/473

ghaf on HEAD (139d614) [$] via ❄️  impure (Ghaf-devshell-env)
❯ nix build .#packages.x86_64-linux.lenovo-x1-carbon-gen11-debug

## reference - origin/main

ghaf on HEAD (2db4cee) [$] via ❄️  impure (Ghaf-devshell-env)
❯ nix build .#packages.x86_64-linux.lenovo-x1-carbon-gen11-debug

## Test protocol - load on Youtube Full HD streaming

* Connect wifi network
* Open chromium-vm with https://www.youtube.com/watch?v=RzVvThhjAKw
* Quality > 1080p
* Play full screen
* Right click > Stats for nerds
* Open terminal1	> htop (gui-vm)
* Open terminal2 > ssh 192.168.101.2 > htop (ghaf-host)
* Open terminal3 > ssh chromium-vm > htop (chromium-vm)
* Place terminal windows on top of full screen video (see images below)
* Take an image with camera

## Results

### https://github.com/tiiuae/ghaf/pull/473 @ 139d614
![shmem](https://github.com/user-attachments/assets/b833bc79-a69f-4d00-812d-73edf46f09af)

### main @ 2db4cee

![non_shmem](https://github.com/user-attachments/assets/47c4da43-0f94-4220-8113-e412086e8d8c)

## Conclusions

Snapshot of subsystem load numbers (incl. average) are inconclusive as they are ~ in the same ballpark.

vilvo commented on modules/desktop/profiles/applications.nix:
```diff
@@ -4,21 +4,112 @@
+            Defines shared memory size
+          '';
+        };
+        vmCount = mkOption {
```
I have some concerns with `vmCount` used as a duplicate limit for already declared number of application VMs.
If/when we must(?) set this as [a build-time parameter to memsocket](https://github.com/tiiuae/ghaf/pull/473/files#diff-4cd0792a1806df6e4de8a4de75cf451ec4536470b07fa5b4a3f10ad041ed989eR31), could we instead [reuse the length of `enabled-app-vms` list](https://github.com/tiiuae/ghaf/blob/2db4ceeff045a8bc4c858bb1ff33481f70ed7c26/modules/reference/appvms/default.nix#L22) to follow DRY-principle?

As a practical question - what happens if the the `vmCount` is declared smaller than the amount of system enabled application VMs?

jkuro-tii replied:
> As a practical question - what happens if the the `vmCount` is declared smaller than the amount of system enabled application VMs?
1. Interrupts from excessive VMs will not interrupt the peer as it allocates only the `vmCount` pairs of interrupts.
2. The `memsocket` application will fail as there is insufficient shared memory for excessive VMs.

jkuro-tii replied:
> [reuse the length of enabled-app-vms list](https://github.com/tiiuae/ghaf/blob/2db4ceeff045a8bc4c858bb1ff33481f70ed7c26/modules/reference/appvms/default.nix#L22) to follow DRY-principle?

Done.

vilvo commented on modules/desktop/profiles/applications.nix:
```diff
@@ -4,21 +4,112 @@
+        };
+        serverSocketPath = mkOption {
+          type = lib.types.str;
+          default = "/run/user/1000/memsocket-server.sock";
```
Binding the path to `ghaf` `UID` can be error-prone. It might be an acceptable risk but the user-model is due to get changed to operator and admin model where we can't assume the operator user to have the same rights as it currently has. Depending on how nix creates the declared operator and admin users, the UID bound path can break things.

jkuro-tii replied:
I added the `uid` parameter to `options.ghaf.users.accounts`, as the same issue exists for nm-launcher.

vilvo commented on modules/desktop/profiles/applications.nix:
```diff
@@ -4,21 +4,112 @@
+        };
+        clientSocketPath = mkOption {
+          type = lib.types.str;
+          default = "/run/user/1000/memsocket-client.sock";
```
https://github.com/tiiuae/ghaf/pull/473/files#r1698205321

vilvo commented on modules/desktop/profiles/applications.nix:
```diff
@@ -4,21 +4,112 @@
+          default = "0x920000000";
+          description = mdDoc ''
+            If set to a non-zero value, it maps the shared memory
+            into this physical address.
```
Would you please add to `description` on how the default value was/is chosen?

jkuro-tii replied:
Done.

vilvo commented on modules/hardware/common/kernel.nix:
```diff
@@ -50,11 +50,21 @@ in {
           ++ config.ghaf.hardware.definition.gpu.pciDevices
           ++ config.ghaf.hardware.definition.audio.pciDevices
         ));
+        hugepagesz = 2; # MBytes
```
if you use "optional scale suffix [kKmMgG]" here, the comment gets redundant and you can also change the line https://github.com/tiiuae/ghaf/pull/473/files#diff-7e202e79cee502470d29582cd4b072d36760f5cdc8c1eca7e87026577f04ae03R58 where you already use the scale suffix. Changing this also removes a source of error - if one wants to test different page sizes, they can't define the suffix on line 53 because it would break on line 58 with two suffices. Defining it only on line 53 solves the problem.

jkuro-tii replied:
Fixed

vilvo commented on modules/microvm/virtualization/microvm/appvm.nix:
```diff
@@ -25,6 +25,11 @@
       then vm.cid
       else cfg.vsockBaseCID + index;
+    shmConfig = configHost.ghaf.profiles.applications.ivShMemServer;
+    memsocket = pkgs.callPackage ../../../../packages/memsocket {
```
memsocket should be thoroughly reviewed and tested before integrating this PR

vilvo commented on modules/microvm/virtualization/microvm/guivm.nix:
```diff
@@ -227,20 +267,52 @@ in {
+              echo Erasing ${shmConfig.hostSocketPath} ${pidFilePath}
+              rm -f ${shmConfig.hostSocketPath}
+            fi
+            ${pkgs.sudo}/sbin/sudo -u microvm -g kvm ${pkgs.qemu_kvm}/bin/ivshmem-server -p ${pidFilePath} -n ${vectors} -m /dev/hugepages/ -l ${(toString config.ghaf.profiles.applications.ivShMemServer.memSize) + "M"}
```
Please avoid redefining the suffix `"M"` here.

jkuro-tii replied:
It's not a redefinition. I added a comment that defined shared memory size unit are MBytes:
https://github.com/jkuro-tii/ghaf/blob/4717c4e9df348dbde0e336a273365fa050daa440/modules/desktop/profiles/applications.nix#L27

vilvo commented on overlays/custom-packages/qemu/0001-ivshmem-flat-memory-support.patch:
Have you asked QEMU upstream for comments on this patch?

jkuro-tii replied:
Not yet. I plan to do it after finishing this review.

vilvo reviewed (changes requested):
In general, this is a great step forward on functional shared memory support on an area where it can matter a lot. My testing was not conclusive on performance improvement https://github.com/tiiuae/ghaf/pull/473#issuecomment-2259930145 but I recommend involving testing team on through regression testing and further performance testing.

The main change request is to get proper peer review for `memsocket` before integrating this PR.
Smaller change requests are added in detailed comments. 
Thanks for addressing previous review comments and improving the PR description. All in all, nice work. A lot of potential for perf improvements, optimization and security (potentially with pKVM).

avnik commented on packages/memsocket/default.nix:
```diff
@@ -0,0 +1,49 @@
+      else ""
+    );
+  prePatch = ''
+    cd app
```
If app not in repo root, but in `app/` subdirectory -- use `sourceRoot="app/"` statement, not chdir during prePatch phase

jkuro-tii replied:
Done. Surprisingly, the `sourceRoot` must be headed with `source`, i.e.: `sourceRoot="source/app/"`.

avnik commented on packages/memsocket/memtest.nix:
```diff
@@ -0,0 +1,39 @@
+    sha256 = "sha256-uaKBtb+bD42RK/Z96dhMrGxNXg+ZV/6iW52ouNQ+QSY=";
+  };
+
+  nativeBuildInputs = [gcc gnumake];
```
GCC and gnumake is in stdenv, no need to set them explicitly. Also I suggest to join memsocket and memtest to single package (just add there root makefile to make/make install all at once.

jkuro-tii replied:
I decided to remove the memtest component. It's used only for checking the speed and reliability of the socket-to-socket connection.

vilvo commented:
> The main change request is to get proper peer review for `memsocket` before integrating this PR.

I propose (at least) static analyser protocol and fixes to `memsocket` (a.k.a `shmsockproxy`):

```
$ git clone https://github.com/tiiuae/shmsockproxy.git
$ cd shmsockproxy/
$ nix-shell -p flawfinder
$ flawfinder .
...
ANALYSIS SUMMARY:

Hits = 43
Lines analyzed = 4110 in approximately 0.02 seconds (190030 lines/second)
Physical Source Lines of Code (SLOC) = 3207
Hits@level = [0]  17 [1]  10 [2]  18 [3]   0 [4]  15 [5]   0
Hits@level+ = [0+]  60 [1+]  43 [2+]  33 [3+]  15 [4+]  15 [5+]   0
Hits/KSLOC@level+ = [0+] 18.7091 [1+] 13.4082 [2+] 10.29 [3+] 4.67727 [4+] 4.67727 [5+]   0
Dot directories skipped = 1 (--followdotdir overrides)
Minimum risk level = 1

Not every hit is necessarily a security vulnerability.
You can inhibit a report by adding a comment in this form:
// flawfinder: ignore
Make *sure* it's a false positive!
You can use the option --neverignore to show these.
$ flawfinder . > results.txt
```
Please find the results attached.
[results.txt](https://github.com/user-attachments/files/16452262/results.txt)

vunnyso commented on modules/microvm/virtualization/microvm/appvm.nix:
```diff
@@ -42,10 +47,18 @@
+          runWaypipe =
+            if shmConfig.display
+            then
+              pkgs.writeScriptBin "run-waypipe" ''
```
Is it possible to have `if, else` in script rather creating than creating different scripts?

jkuro-tii commented:
Updates:

- Refactored Nix code by consolidating most host and VM configuration changes into a single file, improving maintainability.

- Added support for shared memory on the host system, enabling any application utilizing Unix Sockets (e.g. gRPC) to seamlessly run on both the host and within a virtual machine (VM).

- Implemented the shared memory driver in the NixOS build as a kernel module instead of a kernel patch, simplifying maintenance and eliminating the need for kernel rebuilding.

- Replaced the hardcoded ghaf user ID (1000) with a configurable parameter, increasing flexibility.

UnikieHR commented:
![image](https://github.com/user-attachments/assets/51b93d5e-52e5-474e-b313-fb927a7b0d66)

mbssrc commented on modules/hardware/common/shared-mem.nix:
```diff
@@ -0,0 +1,232 @@
+  pkgs,
+  ...
+}:
+with lib;
```
Common pattern used atm is 
``` 
: let
  inherit (lib) types; # + all the functions you need
in {
```

mbssrc commented on modules/hardware/common/shared-mem.nix:
```diff
@@ -0,0 +1,232 @@
+with lib;
+{
+  options.ghaf.shm = {
+    enable = lib.mkOption {
```
and no need for .lib

mbssrc commented on modules/hardware/common/shared-mem.nix:
```diff
@@ -0,0 +1,232 @@
+    };
+  };
+
+  config.boot.kernelParams =
```
For better visibility in the module, you can write 
```
config = mkIf cfg.enable {
  # all configurations here, e.g.,
  boot.kernelParams = ...
}
```

mbssrc commented on modules/hardware/common/shared-mem.nix:
```diff
@@ -0,0 +1,232 @@
+        else
+          enable;
+    };
+  };
```
Do you need all of these options to be configurable, or are they referenced outside of the module? If not, you can define them in the `let ... in` part. Also, you can define assertions in the config if needed

mbssrc commented on modules/hardware/common/shared-mem.nix:
```diff
@@ -0,0 +1,232 @@
+    let
+      hugepagesz = "2M"; # valid values: "2M" and "1G", as kernel supports these huge pages' size
+      hugepages =
+        if hugepagesz == "2M" then config.ghaf.shm.memSize / 2 else config.ghaf.shm.memSize / 1024;
```
Should `hugepagez = "2M";` be an option? Otherwise the `hugepages` calculation is always the same

mbssrc commented on modules/hardware/common/shared-mem.nix:
```diff
@@ -0,0 +1,232 @@
+            echo Erasing ${config.ghaf.shm.hostSocketPath} ${pidFilePath}
+            rm -f ${config.ghaf.shm.hostSocketPath}
+          fi
+          ${pkgs.sudo}/sbin/sudo -u microvm -g kvm ${pkgs.qemu_kvm}/bin/ivshmem-server -p ${pidFilePath} -n ${vectors} -m /dev/hugepages/ -l ${
```
Could you run the systemd service as `User = "microvm"; Group = "kvm";` instead of using sudo?

jkuro-tii replied:
Running the service under root is needed for setting ownership of /dev/hugepages - it must be done under `root`. Elevating privileges in a user-owned service using /run/wrappers/bin/sudo is not feasible since it prompts for a password. Additionally, using ${pkgs.sudo}/sbin/sudo is also not an option in this case.

mbssrc commented on modules/hardware/common/shared-mem.nix:
```diff
@@ -0,0 +1,232 @@
+
+  config.ghaf.hardware.definition.gpu.kernelConfig.kernelParams = optionals config.ghaf.shm.enable [
+    "kvm_ivshmem.flataddr=${config.ghaf.shm.flataddr}"
+  ];
```
Could you use `config.microvm.vms.gui-vm.config.config.boot.kernelParams` directly instead of the hardware configuration?

brianmcgillion commented:
@jkuro-tii can you squash those commits to allow easier bisect?

leivos-unikie commented:
I tried to run the tests described in 
https://ssrc.atlassian.net/wiki/spaces/~62552e6ffdb60b006927ad98/pages/825720835/Wayland+displaying+with+shared+memory
but I think I have missed something.

```
[ghaf@gui-vm:~]$ systemctl start --user memsocket.service
Failed to start memsocket.service: Unit memsocket.service not found.
```
```
[ghaf@chromium-vm:~]$ systemctl stop --user memsocket.service
Failed to stop memsocket.service: Unit memsocket.service not loaded.
```
```
[ghaf@ghaf-host:~]$ vsockproxy 5201 100 5201
vsockproxy: command not found
```

leivos-unikie commented:
And also I am wondering if those Confluence pages under Documentation have been migrated to new tii hosted Confluence.

leivos-unikie commented:
tests in ci-test-automation passed

leivos-unikie commented:
Since the shared memory feature is disabled by default (for now) I ran the iperf tests described in documentation with separate testing branch
https://github.com/jkuro-tii/ghaf/tree/jkuro-add-wayland-shm-lenovo_rebase_test

**In gui-vm:**
`systemctl stop --user memsocket.service`
`memsocket -s 1234.sock 3 &`
`UNSOCK_ADDR=127.0.0.1 UNSOCK_DIR=``pwd`` LD_PRELOAD=/nix/store/dlavg00s1ckmkdkcaz1p9igqnmzgnhf6-unsock/lib/libunsock.so.1.1.0 iperf -c 127.0.0.1 -p 1234`

**In chromium:**
`systemctl stop --user memsocket.service`
`UNSOCK_BLOCK_INET6=1 UNSOCK_PORT=55555 UNSOCK_ADDR=127.0.0.1/0 UNSOCK_DIR="/home/ghaf/" LD_PRELOAD=/nix/store/dlavg00s1ckmkdkcaz1p9igqnmzgnhf6-unsock/lib/libunsock.so.1.1.0 iperf -s -p 55555 &`

-> Speed 20.1Gbits/sec


For comparison without shared memory

**In gui-vm:**
`iperf -s -1 -p 5201`

**In chromium-vm:**
`iperf3 -c 192.168.100.3 -p 5201`

-> Speed 15Gbits/sec

leivos-unikie commented:
Should there be some documentation about this feature in ghaf/docs?
(Jenni had already asked this but didn't receive any answer.)

At least in my opinion it is not enough to have the iperf test instructions in the old Confluence.

vadika commented:
> Should there be some documentation about this feature in ghaf/docs? (Jenni had already asked this but didn't receive any answer.)
> 
> At least in my opinion it is not enough to have the iperf test instructions in the old Confluence.

Documentation is definitely needs to be in place in the Documentation repository
